core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java

```diff
@@ -1088,9 +1091,32 @@ public NamespaceMapping getNamespaces() {
     return namespaces;
   }
 
-  public HashMap<TableId,ClientTabletCache> tabletCaches() {
+  public ClientTabletCache getTabletLocationCache(TableId tableId) {
     ensureOpen();
-    return tabletCaches;
+    synchronized (tabletLocationCache) {
+      return tabletLocationCache.computeIfAbsent(tableId, (TableId key) -> {
+        var mlo = new MetadataCachedTabletObtainer();
+        var lockChecker = getTServerLockChecker();
+        if (AccumuloTable.ROOT.tableId().equals(tableId)) {
+          return new RootClientTabletCache(lockChecker);
+        } else if (AccumuloTable.METADATA.tableId().equals(tableId)) {
+          return new ClientTabletCacheImpl(AccumuloTable.METADATA.tableId(),
+              getTabletLocationCache(AccumuloTable.ROOT.tableId()), mlo, lockChecker);
```

Comment:
Inside computeIfAbset this function will call computeIfAbsent again w/ a different key which you are not supposed to do.  If the cache were per datalevel, then this would be safe (like if the top level cache was `Map<DataLevel, <Map<TableId,ClientTabletCache>>`